### ISSUE TITLE
Cassandra single table inheritance issue

### ISSUE THREAD

**Issue by uladzimirFilipchanka** [Impetus/kundera#687]
Hello,

We've found strange behaviour with JPA model inheritence (SINGLE_TABLE mapping) on insert.
Looks like kundera tries to add two mention of column which is descriminator: first time as simple column from metadata, second time as descriminator column

We're using 2.14 kunderra + jboss infrastructure.
Here is our hierarchy (names changed a bit, getters/setters omitted)

``` java
@Entity
@Table(name = "animal")
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "type", discriminatorType = DiscriminatorType.INTEGER)
public abstract class Animal implements Serializable {
    @Id
    @GeneratedValue
    @Column(name = "id")
    private String id;

    @Column(name = "type")
    private int type;
}

@Entity
@DiscriminatorValue(value = "1")
public class Dog extends Animal {
    @Column(name = "height")
    private int height;
}

@Entity
@DiscriminatorValue(value = "2")
public class Cat extends Animal {
    @Column(name = "name")
    private String name;
}
```

And DAO looks very straightforward:

``` java
public void saveDog(Dog dog)
{
    entityManager.persist(dog);
    entityManager.flush();
}
```

After transaction commit we've got:
**Error while executing query  INSERT INTO "animal"("type","height","id","type") VALUES(0,12,'a7b71b50-973a-11e4-b8b3-f766e78c0682','2')**

Meaningful part of stacktrace:

```
Caused by: com.impetus.kundera.KunderaException: com.datastax.driver.core.exceptions.InvalidQueryException: Multiple definitions found for column type
 at com.impetus.kundera.client.cassandra.dsdriver.DSClient.execute(DSClient.java:510) [kundera-cassandra-ds-driver-2.14.jar:]
 at com.impetus.client.cassandra.CassandraClientBase$CQLClient.persist(CassandraClientBase.java:1826) [kundera-cassandra-2.14.jar:]
 at com.impetus.kundera.client.cassandra.dsdriver.DSClient.onPersist(DSClient.java:112) [kundera-cassandra-ds-driver-2.14.jar:]
 at com.impetus.kundera.client.ClientBase.persist(ClientBase.java:105) [kundera-core-2.14.jar:]
 at com.impetus.kundera.lifecycle.states.ManagedState.handleFlush(ManagedState.java:191) [kundera-core-2.14.jar:]
 at com.impetus.kundera.graph.Node.flush(Node.java:531) [kundera-core-2.14.jar:]
 at com.impetus.kundera.persistence.DefaultTransactionResource.onFlush(DefaultTransactionResource.java:88) [kundera-core-2.14.jar:]
 at com.impetus.kundera.persistence.DefaultTransactionResource.onCommit(DefaultTransactionResource.java:72) [kundera-core-2.14.jar:]
 at com.impetus.kundera.persistence.Coordinator.coordinate(Coordinator.java:79) [kundera-core-2.14.jar:]
 at com.impetus.kundera.persistence.KunderaEntityTransaction.onTransaction(KunderaEntityTransaction.java:128) [kundera-core-2.14.jar:]
 at com.impetus.kundera.persistence.KunderaEntityTransaction.commit(KunderaEntityTransaction.java:78) [kundera-core-2.14.jar:]
 at org.ustaxi.app.infra.cassandra.CassandraTxIntercepter.wave(CassandraTxIntercepter.java:49) [infra-cassandra-0.2.jar:]
 at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) [rt.jar:1.7.0_71]
 at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57) [rt.jar:1.7.0_71]
 at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) [rt.jar:1.7.0_71]
 at java.lang.reflect.Method.invoke(Method.java:606) [rt.jar:1.7.0_71]
 at org.jboss.weld.interceptor.proxy.SimpleMethodInvocation.invoke(SimpleMethodInvocation.java:30) [weld-core-impl-2.1.2.Final.jar:2014-01-09 09:23]
 at org.jboss.weld.interceptor.chain.AbstractInterceptionChain.invokeNext(AbstractInterceptionChain.java:103) [weld-core-impl-2.1.2.Final.jar:2014-01-09 09:23]
 at org.jboss.weld.interceptor.chain.AbstractInterceptionChain.invokeNextInterceptor(AbstractInterceptionChain.java:81) [weld-core-impl-2.1.2.Final.jar:2014-01-09 09:23]
 at org.jboss.weld.interceptor.proxy.InterceptorMethodHandler.executeInterception(InterceptorMethodHandler.java:48) [weld-core-impl-2.1.2.Final.jar:2014-01-09 09:23]
 at org.jboss.weld.interceptor.proxy.InterceptorMethodHandler.invoke(InterceptorMethodHandler.java:41) [weld-core-impl-2.1.2.Final.jar:2014-01-09 09:23]
 at org.jboss.weld.bean.proxy.CombinedInterceptorAndDecoratorStackMethodHandler.invoke(CombinedInterceptorAndDecoratorStackMethodHandler.java:53) [weld-core-impl-2.1.2.Final.jar:2014-01-09 09:23]
```

After debugging I've found that first append of 'type' column to resulting query happens at `com.impetus.client.cassandra.thrift.CQLTRanslator#onTranslation(231)`. Somewhere in while loop. As far as I understand, this time kundera takes the column as 'regular' one, from metadata. And the second append happens at `om.impetus.client.cassandra.thrift.CQLTranslator#onDiscriminatorColumn(329)`. The framework takes the column as descriminator column.

 And the second short point. 
 We've tried to marked 'type' as `@Transient`. In this we have problems duging entityManager.find(). Kundera allows both calls:

``` java
 entityManager.find(Dog.class, '1')
 entityManager.find(Cat.class, '1')
```

 So looks like without actual column in JPA the framework doesn't take descriminator into account.

**Comment by pragalbhgarg** [comment]
@uladzimirFilipchanka 
You need not add discriminator column specifically as a separate entity attribute i.e. this should be removed

```
@Column(name = "type")
    private int type;
```

Hope that helps !
-pragalbh

**Comment by mevivs** [comment]
:+1:  for @pragalbhgarg  that is the reason behind multiple definitions found.

-Vivek

**Comment by uladzimirFilipchanka** [comment]
Thanks for the response.
Yes, I've tried removing 'type' column from base class at all. But I've got exception during selecting data. I have following code for getCatByName in DAO object (omitting non-unique issues):

```
Query query = getEM().createQuery("select c from Cat c where c.name =:name");
query.setParameter("name", name);
return (Cat) query.getSingleResult();
```

   Exception throws on third line: 'java.lang.IllegalArgumentException' exception.

```
at com.impetus.kundera.metadata.model.type.AbstractManagedType.checkForValid(AbstractManagedType.java:1019)
at com.impetus.kundera.metadata.model.type.AbstractManagedType.getAttribute(AbstractManagedType.java:510)
at com.impetus.kundera.metadata.model.type.AbstractManagedType.getAttribute(AbstractManagedType.java:507)
at com.impetus.kundera.metadata.model.MetamodelImpl.getEntityAttribute(MetamodelImpl.java:372)
at com.impetus.client.cassandra.query.CassQuery.onCondition(CassQuery.java:790)
at com.impetus.client.cassandra.query.CassQuery.onQueryOverCQL3(CassQuery.java:687)
at com.impetus.client.cassandra.query.CassQuery.populateEntities(CassQuery.java:136)
at com.impetus.kundera.query.QueryImpl.fetch(QueryImpl.java:1013)
at com.impetus.kundera.query.QueryImpl.getResultList(QueryImpl.java:164)
at com.impetus.kundera.query.QueryImpl.getSingleResult(QueryImpl.java:442)
 ...
```

attribute of the given name and type is not present in the managed type, for name:null

**Comment by mevivs** [comment]
What's

```
 c.name
```

Can you please point me to trace it in entity definition?

-Vivek

**Comment by mevivs** [comment]
@uladzimirFilipchanka 
My bad i got it! Please share updated entity definitions.

-Vivek

**Comment by uladzimirFilipchanka** [comment]
@mevivs
Entities stay the same. 

```
c.name
```

is field in Cat class. In general, any query against Cat or Dog entities cause IllegalArgumentException.

**Comment by chhavigangwal** [comment]
Fixed. Released with Kundera-2.15. Can you please verify at your end ?

Chhavi

**Comment by uladzimirFilipchanka** [comment]
@chhavigangwal 
Yep, it's work. Thanks a lot!
